packages/core/src/hooks/useQueryParamState.ts

```diff
@@ -83,10 +81,16 @@ export function useQueryParamState<T>(
       );
 
       if (location.search !== queryString) {
-        navigate({ ...location, search: `?${queryString}` }, { replace: true });
+        // We fallback to the history API, as navigate from react-router causes
+        // input elements to loose focus.
+        history.replaceState(
```

Comment:
Oof, this is window.history directly isn't it? Was there really no set of parameters to useNavigate that would not lead to this problem?

Reply:
My concern is whether this really plays well with react-router, i.e. does it break proper back/forward navigation or something like that, in edge cases

Reply:
Yeah, this is the plain history API. I couldn't find any issues, but that doesn't mean there aren't any. The problem is that this also assumes that this runs in a browser, so using this in a memory router shouldn't work. That could cause issues in tests. react-router also comes with a `useSearchParams` with is easier to use than `useNavigate`, but has the same issue.
Underneath the `history` package is used, but I can't access it.

I assume that the code with `useSearchParams` should work and will check whether the issue is somewhere else in the code.

Reply:
Looks like `useSearchParams` works fine when I use it for an input outside of the api docs page. So I guess it will be something about the table -.-.

Reply:
Yep or the input or part of the form parents (implicitly?) depending on the location in such a way that a repaint is forced when it should not be. I bet that's what happens, maybe it's not the history api as such, but rather that something is driven by the location not just on page load but after that as well. Maybe a larger search for useLocation etc in the vicinity of these components could be informative

Reply:
So the main problem is, that https://github.com/backstage/backstage/blob/master/packages/core/src/components/Table/Table.tsx#L382-L431 is recreated too often. It goes away once I remove the `toolbarClasses` and `filterClasses` from the dependency array. It still is the same if I just add `toolbarClasses.root` & co which I think is quite strange 😆 I will refactor it a bit so that it doesn't rerender that often anymore.

Reply:
Nice! Yeah suspicion confirmed then.

Reply:
@freben Problems solved. I changed a bit in the `Table` itself and tried to improve the dependencies there. now it works. I also switched to `useSearchParams` as it's easier to use.